frontend/catalyst/pennylane_extensions.py

```diff
@@ -50,6 +50,44 @@
 # pylint: disable=too-many-lines
 
 
+def _trace_quantum_tape(
+    cargs, ckwargs, qargs, _callee: Callable, _allow_quantum_measurements: bool = True
+) -> Tuple[Any, Any]:
+    """Jax-trace the ``_callee`` function accepting positional and keyword arguments and containig
+    quantum calls by running it under the PennyLane's quantum tape recorder.
+
+    Args:
+        cargs (Jaxpr): classical positional arguemnts to be passed to ``_callee``
+        kwargs (Jaxpr): classical keyword arguemnts to be passed to ``_callee``
+        qargs (Jaxpr): quantum arguments to consume in the course of tracing
+        _callee (Callable): function to trace
+        _allow_quantum_measurements (bool): If set to False, raise an exception if quantum
+                                            measurements are detected
+    Returns (Tuple[Any,Any]):
+        - Jax representaion of classical return values of ``_callee``
+        - Jax representation of quantum return values obtained in the course of tracing
+    """
+    assert len(qargs) == 1, f"A single quantum argument was expected, got {qargs}"
+    with qml.QueuingManager.stop_recording():
+        with JaxTape() as tape:
+            with tape.quantum_tape:
+                out = _callee(*cargs, **ckwargs)
+            if not _allow_quantum_measurements and len(tape.quantum_tape.measurements) > 0:
+                raise ValueError("Quantum measurements are not allowed in this scope")
+            if isinstance(out, Operation):
+                out = None
+            tape.set_return_val(out)
+            new_quantum_tape = JaxTape.device.expand_fn(tape.quantum_tape)
+            tape.quantum_tape = new_quantum_tape
+            tape.quantum_tape.jax_tape = tape
+
+    has_tracer_return_values = out is not None
+    qreg = qargs[0]
+    return_values, qreg, qubit_states = trace_quantum_tape(tape, qreg, has_tracer_return_values)
+    qreg = insert_to_qreg(qubit_states, qreg)
+    return return_values, [qreg]
```

Comment:
Why do we need qreg in a list?

Reply:
It turned out that we use a single register everywhere right now, but this leads to a disadvantageous requirement of inserting extract/insert too often. Here I am proposing to use a slightly more generic interface so we can have a room for improvement.